### ISSUE TITLE
git clean removes content of symlinks

### ISSUE THREAD

**Issue by illuxio** [git-for-windows/git#607]
Hallo,

I'm using git version 2.7.0. I discovered recently that a `git clean -xdf`deletes the content of my symlinks that are in my `.gitignore` file instead of removing the symlinks as in unix.

The symlinks was created by the command `npm link`.

I hop you will fix this soon.

**Comment by dscho** [comment]
> The symlinks was created by the command npm link.

Please note that it is unreasonable to expect everybody to have `npm` installed (besides, you did not specify which `npm`. So you need to do a little homework and figure out what `npm` calls underneath, and come up with a [Minimal, Complete & Verifiable Example](http://stackoverflow.com/help/mcve).

> I hop you will fix this soon.

I hope you realize that a statement like this is prone to cause the opposite of what you want. After all, who likes to be underappreciated *and* pushed around?

**Comment by illuxio** [comment]
As far as I understand npm uses junctions (can be created by `mklink \J Link Target`).

To reproduce the bug:
- create git repo with files
   * .gitignore
   * target/README.md
- Add `/link` to `.gitignore`
- Commit all files
- Execute in cmd.exe `mklink \J link target` to link the link and target
- Execute `git clean -xdf`

Now you will see that README.md is missing in target.

**Comment by rimrul** [comment]
>* create git repo with files
* .gitignore
* target/README.md
* Add /link to .gitignore
* Commit all files
* Execute in cmd.exe mklink \J link target to link the link and target
* Execute git clean -xdf

Ummm... you have the folder target inside the repository, add the files inside it to the repository, and wonder that `git clean -xdf` doesn't ignore the folder?

**Comment by rimrul** [comment]
>* create git repo with files
* .gitignore
* target/README.md
* Add /link to .gitignore
* Commit all files
* Execute in cmd.exe mklink \J link target to link the link and target
* Execute git clean -xdf

Ummm... you have the folder target inside the repository, add the files inside it to the repository, and wonder that `git clean -xdf` doesn't ignore the folder? Does this only seem like the intended behaviour to me?

**Comment by illuxio** [comment]
I don't think so, because executing `rm -rf link` in the git bash or deleting the link in the explorer will not result in an empty target folder...

**Comment by rimrul** [comment]
> I don't think so, because executing rm -rf link in the git bash or deleting the link in the explorer will not result in an empty target folder...

How about `rm -rf target`?

**Comment by illuxio** [comment]
Sure this will delete the target folder and the README.md, but this is not the point.
For me sound this like the question why `git clean -xdf` does not remove all file from the git repo.

My problem is that I have a folder "target" with files and obviously a junction "link" (a type of symlink) that links to this folder. Target and is files is wanted inside the repo and the junction "link" is not. Therefore, the the junction "link" is in the `.gitignore` file.

What I expected from `git clean -xdf` is that all wanted files style and all unwanted files are removed. This is not the case because "link" and all files inside "target" are removed.

**Comment by rimrul** [comment]
>Sure this will delete the target folder and the README.md, but this is not the point.
For me sound this like the question why git clean -xdf does not remove all file from the git repo.

Sorry. I messed this up by not looking up, what exactly `git clean -xdf` is supposed to do...

Please accept my appology for wasting your time. 

I can confirm that `git clean -xdf` doesn't touch the file on a Linux symlink.

**Comment by illuxio** [comment]
No problem. I think with my addition it is now more precise.

**Comment by whoisj** [comment]
so basically you want `git clean -xdf` to not follow junctions, but delete them unless they're ignored with a `-e` flag?

**Comment by illuxio** [comment]
Exactly.

J Wyman <notifications@github.com> schrieb am Mi., 20. Jan. 2016 19:19:

**Comment by whoisj** [comment]
> Exactly.

I implemented similar logic for Libgit2 some time ago. I can assist here as well, though I doubt I'd have time to generate a full pull request on the topic.

Win32's FindFirst/Next iterator structure returns enough information in `WIN32_FIND_DATA` to detect if a node is a reparse point or not (`(fd.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)`. From there a `DeviceIoControl` with `FSCTL_DELETE_REPARSE_POINT` should achieve what you want - the important piece is to detect the reparse point (aka Junction) and handle it as one.

**Comment by dscho** [comment]
After such a long time, I have made up my mind. As pointed out in https://github.com/git-for-windows/git/pull/437, junction points are just different enough from symlinks to *not* be treated as such. If you create a directory junction point, you really make the target directory accessible as a directory in the local directory. Note: this is *not* a symbolic link, but more like a "hard link" (if directories could be hard linked).

In short: if you have a directory junction point and try to delete it recursively, I think that we do the exact right thing by removing the directory *contents*, too.

@illuxio if you disagree, please try to convince me that this is an improper take. And be prepared to do the heavy-lifting of the change after that.

**Comment by stasberkov** [comment]
@dscho *git* for windows shall handle *junctions* the same way as other Windows programs do -- delete junction but not its content. I.e. when in *Windows Explorer* you delete directory that contains *junction* you delete just junction but *not* its content. The same way other Windows file managers behave. E.g. cmd.exe command `rmdir dir-that-have-inner-junction /s` removes just *junction* but not junction content even `/s` recursive option is specified. Far Manager behaves the same way.

**Comment by dscho** [comment]
@stasberkov I could be convinced by a patch implementing that logic ;-)

**Comment by hypersw** [comment]
Directory junctions are the most convenient softlinks for Windows NT directories — the symlink directory softlinks have lots of problems, like requiring special privileges, being interpreted at user level rather than by filesystem, which results in different processing depending on the client (e.g. when exposing the folder as a network share). So junctions are a sane choice. And we usually expect them to behave as softlinks in case of deletion.

We experience undesired side-effects from the `clean` command on softlinks done thru junctions. On client machines, this corrupts files shared between working copies in the context of the other working copy. On server machines, this unwantedly wipes up cached files.

How about the implementation which isn't dependent on extra system calls (we use similar technique in ReSharper and Rider filesystem components to handle deletion consistently):
* First, delete the directory filesystem object itself.
  * On an empty directory, this just deletes. All done.
  * On a softlink source, this breaks the link and deletes the source directory. The target remains untouched. All done.
  * On a real non-empty directory, this does nothing and returns a refuse result. Ignore this and proceed.
* Second (unless done), recurse into child directories and files to run the full delete for them.
* Third, delete the directory again. This time an error is an error.

The second and third steps are already there, this just adds the first probe to kill the directory (or link source) directly. Haven't seen any noticeable side-effects from this so far. Is very easy to implement. Does not dig into filesystem implementation details.
What do you think?

**Comment by whoisj** [comment]
@hypersw I like your approach, it makes a lot of sense; however I have to say that there are already those that depend on the existing behavior and changing it could/would cause them heartburn and send them here complaining in the opposite direction.

The best solution here would be for somebody who really wants and/or needs this functionality to offer up a solution via pull-requests which implements the solution; and places it behind a configuration flag so that users who depend on the existing functionality aren't broken by the improved functionality.

Why do I suggest "somebody else" do the work? Personally, I'm already 110% scheduled and @dscho is likely in a similar situation. If those who want/need this change are happy waiting until such a time that somebody has available time to implement it, then that's fine - however if they're not happy to wait, then they need to act. Open-source is awesome like that. 😄

**Comment by PhilipOakley** [comment]
@hypersw Is the idea easy to code and test within the GfW codebase? A proof of concept (PoC) pull request (PR) may help things along. 

Does this make all the mechanisms (junctions, reparse points, etc) the same as the upstream Git's view of symlinks? And if not, how should it be documented so that it is really easy to understand (especially as 95% of web explanations would be the upstream viewpoint, so Windows users may never see the documentation...).

**Comment by BurtHarris** [comment]
Running into the same problem in the context of building a mulit-package repro with junctions set up by the NPM package manager.  The behavior of `git clean -idx` is extremely surprising and can be difficult for someone without a lot of git experience to recover from.   @hypersw's proposed solution makes a lot of sense to me.   

I could find the time to try to tackle this, but I will probably need some help navigating this repro's conventions.    

@whoisj I have a hard time imagining how anyone might consider this behavior a feature.   Can you cite an example?

**Comment by BurtHarris** [comment]
In tests (*** t1013-read-tree-submodule.sh ***) I'm getting messages like `
not ok 4 - git read-tree -u -m --recurse-submodules: added submodule removes an untracked ignored file # TODO known breakage`.   

Does the TODO comment indicate these are known issues I can ignore in testing my PR?

**Comment by halldorfannar** [comment]
@whoisj I have implemented a solution in pull request #1414. I didn't put it behind a configuration flag because what I discovered was that the code base was treating junctions inconsistently. In one out of three places it was being treated as a link but in the other places it was just treating a junction as local directory.  I could put this behind a configuration flag if you insist but would then suggest the default setting would execute this new code path rather than the old.  Please review the pull request at your convenience.

**Comment by BurtHarris** [comment]
My alternative PR is for discussion only.   Anyone who knows the right workflow for this project is welcome to take the lead on it.

**Comment by drizzd** [comment]
I encountered this issue while using the [Node.js package manager (npm)](http://npmjs.com) and [terraform](https://www.terraform.io/). Both use Junctions in Windows to point from a cache directory to another directory inside the same project. The Junction is untracked, whereas the files in the directory it points to are tracked. Here is a more concrete example for npm:

```sh
mkdir lib top
echo '{"name":"lib","version":"0.0.0"}' >lib/package.json
echo '{"name":"top","version":"0.0.0"}' >top/package.json
cd top
npm install file:../lib
```

This will create a junction at `node_modules/lib` pointing to `lib`. When committing the Node.js project to Git, `git clean -qdfx` will delete the tracked and untracked files in the lib directory:

```sh
cd ..
git init
echo node_modules >.gitignore
git add -A
cd top
git clean -qdfx
```

I have lost data this way since git clean would not normally delete files which are not in a subdirectory of the current directory. Googling [why does git clean sometimes delete tracked files](https://stackoverflow.com/questions/52083500/why-does-git-clean-xfd-sometimes-delete-tracked-files-how-can-we-fix-this) indicates that other users are hitting this issue as well.

Git is a tool for working source code, including tools such as npm and terraform. It is a fact that such tools are using junctions as the Windows equivalent of symbolic links. To support this usage, Git clean must not follow the Junction as if it were a mount point.

Maybe this behavior is unexpected and unfortunate for users who would like to use junctions as mount points. However, this is not a primary use case for Git, and at least those users do not suffer unexpected data loss (other than the removal of the mount point). The usefulness of mount points inside Git work trees is limited also on Linux, since Git will give the following error if there is a mount point between the .git directory and the current working directory:

```
fatal: not a git repository (or any of the parent directories): .git
```

Therefore, I strongly vote for merging #437. I am happy to work on improving the patch. But treating the junction as a mount point will not prevent the data loss when junctions are used as symlinks, such as in the npm/terraform scenario.

**Comment by vidartf** [comment]
For now, I'm using an alias instead of clean, that does the following:
```
git add --all
git stash
git clean -qxdf
git clean -qxdf
git clean -qxdf
git clean -qxdf
git clean -qxdf   # Need to run this multiple time to sort out all the symlinks
git reset --hard HEAD  # Restore any files incorrectly deleted by git
git stash pop
```

**Comment by Bart76** [comment]
Had the same issue today. Found out that this can be fixed adequately in (recent versions/updates of) Windows 10. Windows 10's MKLINK command for create symbolic links and junctions is somewhat more sophisticated than the MKLINK command in earlier versions of Windows. My post assumes that a recent version of Windows 10 is used. I don't know anything about the behavior of MKLINK in earlier versions of Windows and I am not planning to test it in earlier versions of Windows either. My apologies for that.

When it comes to directories, MKLINK supports the /D and the /J switch. The former creates a (soft) symbolic link and the latter creates a (hard) junction.

I found out, that:
* when a symbolic link for a directory (using the /D switch of the MKLINK command) is created in a git repository's local folder, it will be deleted when running `git clean -fdx`, but the contents of the link's original folder will remain intact, but
* when a junction for a directory (using the /J switch of the MKLINK command) is created in a git repository's local folder, it will be treated as if it were a real directory and its contents will be deleted too when running `git clean -fdx`.

**Comment by cbuchacher** [comment]
@dscho Thank you for picking up work on this in [b494e36](/dscho/git/commit/b494e36550441f241c69c5c6c1c06fd846be8655). Is there anything I can do to help?

**Comment by dscho** [comment]
@cbuchacher I see you found the PR... ;-)

**Comment by vidartf** [comment]
After having had this problem for a long while now, and having [dealt with similar problems in CPython](https://github.com/python/cpython/pull/5998), I think the correct solution is to:
- treat junctions that are volume mount points as a mount point would on nix.
- treat all other junctions as directory symlinks.

To distinguish between them, compare the output of `GetFullPathNameW` and `GetVolumePathNameW`. If they are equal, it is a mount point.

I'll be happy to help contribute an implementation of this, as long as someone with a better understanding of the current code base is able to point out all the spots where it should be used (I'm guessing the current PR is a good start), and to shepherd the PR to a merge.

**Comment by dscho** [comment]
>     * treat junctions that are volume mount points as a mount point would on nix.

AFAICT this is the exact problem: Git for Windows _does_ do that. It deletes the files in a mount point as if it was a regular directory.

However, junctions are a heck of a lot easier to create on Windows than mount points on Linux/Unix.

And no, we cannot treat _all_ other junction points as symlinks. Only the ones that are explicitly marked as symbolic links.

Which brings us to https://github.com/git-for-windows/git/pull/1976 that does exactly that.

**Comment by vidartf** [comment]
Just to clarify the "notation" as it where:

> And no, we cannot treat all other junction points as symlinks. Only the ones that are explicitly marked as symbolic links.

How do you mark a junction point as a symlink? My understanding is that you have:
- Directory symlinks (`mklink /D`)
- Directory junction points (`mklink /J`)
- Volume mount points (`mountvol`)
- Third party reparse points.

Am I missing something here?

**Comment by dscho** [comment]
> My understanding is that you have:
> 
>     * Directory symlinks (`mklink /D`)
> 
>     * Directory junction points (`mklink /J`)
> 
>     * Volume mount points (`mountvol`)
> 
>     * Third party reparse points.
> 
> 
> Am I missing something here?

What I was referring to were the directory junction points, which are most like Unix bind mounts.

**Comment by cbuchacher** [comment]
@dscho Is #1976 still blocked by the FSCache issue despite my [analysis](https://github.com/git-for-windows/git/pull/1976#issuecomment-471166468) that it should not matter? I can absolutely take care of the merge conflicts or other small changes.

Regarding the FSCache invalidation, I think we should consider it independently of #1976. However, the effort is unclear to me. For starters, my approach would be to write a test which caches an entry, synchronizes the entry across threads, invalidates the entry in one thread, and finally verifies that the entry is (immediately?) invalidated in another thread. I am overthinking this? Maybe, if you have an idea how I can test this feature easily, that would be a good start.

**Comment by dscho** [comment]
>  Is #1976 still blocked by the FSCache issue despite my [analysis](https://github.com/git-for-windows/git/pull/1976#issuecomment-471166468) that it should not matter? I can absolutely take care of the merge conflicts or other small changes.

No, your analysis should have unblocked it, and I simply lack the time to take care of this PR better.

If you have time to resolve the small changes and merge conflicts, that would be very nice. Thank you!

**Comment by vidartf** [comment]
> What I was referring to were the directory junction points, which are most like Unix bind mounts.

"What is was referring to" means this, right?

> Only the ones that are explicitly marked as symbolic links.

If so, I think we are agreed. Sorry for the noise, but happy to have pinged the issue if it causes progress :+1:

**Comment by dscho** [comment]
@vidartf thank you for chiming in. Maybe I can ask you to have a look at https://github.com/git-for-windows/git/pull/1976, too, and leave your comments?

**Comment by vidartf** [comment]
I'm not sure if I'm able to follow the code closely (unfamiliar code base). When you say:

> And no, we cannot treat all other junction points as symlinks. 

which types of junctions/links/mounts would you want to treat in the current way, and how are they discriminated?

**Comment by dscho** [comment]
> > And no, we cannot treat all other junction points as symlinks.
> 
> which types of junctions/links/mounts would you want to treat in the current way, and how are they discriminated?

I made a mistake, I did not want to say "junction points", I meant to say "reparse points".

We have this code: https://github.com/git-for-windows/git/blob/v2.22.0.windows.1/compat/mingw.c#L2724-L2783

As you can see, symlinks have a "[reparse tag](https://github.com/git-for-windows/git/blob/v2.22.0.windows.1/compat/mingw.c#L2754)" which is `IO_REPARSE_TAG_SYMLINK` for symbolic links.

Other reparse point types include `IO_REPARSE_TAG_MOUNT_POINT` (which I mean by "junction point" because `mklink /j` creates them).

**Comment by vidartf** [comment]
Good that we're clearing up the jargon :) Then to clarify the jargon in my question further:

- Both directory junctions by `mklink /D` and volume mounts by `mountvol` will have `IO_REPARSE_TAG_MOUNT_POINT` set.
- I **think** that we want `moutvol` entries to be treated as they currently are, and only apply the new logic to directory junctions (and they can be differentiated by comparing the outputs of `GetFullPathName` and `GetVolumePathName`).
- I don't know how strongly we should try to keep that behavior though, so if there is some cost to doing that in terms of performance, it might not be worthwhile.

Note that I'm saying "**think**" above, as I'm not sure what will be "correct" for git, but I want to be sure that I bring attention to the point so that an informed decision can be made.

**Comment by dscho** [comment]
Addressed by #2268

**Comment by vidartf** [comment]
Thanks for all the work here 🚀

**Comment by sharwell** [comment]
@dscho it looks like there are a bit over 1,000 different commits claiming to fix this issue. The issue was closed more than a year ago, but even in the past day it appears you made 5 new attempts to fix this? I'm not sure what to make of it.

**Comment by dscho** [comment]
> @dscho it looks like there are a bit over 1,000 different commits claiming to fix this issue. The issue was closed more than a year ago, but even in the past day it appears you made 5 new attempts to fix this? I'm not sure what to make of it.

The solution to this riddle is here: https://github.com/git-for-windows/build-extra/blob/3def24f9d192f1123874004b73f56b5af989c4a6/shears.sh#L12

Git for Windows rebases its patches frequently, in so-called "merging-rebases". That's what you see at work here.
